File: examples/example-android/polar-sensor-data-collector/app/src/main/java/com/polar/polarsensordatacollector/ui/graph/EcgDataHolder.kt
package com.polar.polarsensordatacollector.ui.graph

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

const val ECG_BUFFER_SIZE = 3000

object EcgDataHolder {

    data class EcgSample(val voltage: Int)

    data class EcgState(
        val ecgSamples: List<EcgSample> = emptyList(),
        val currentVoltage: Int = 0
    )

    private val _ecgState = MutableStateFlow(EcgState())
    val ecgState: StateFlow<EcgState> = _ecgState.asStateFlow()

    private val ecgSamplesList = ArrayDeque<EcgSample>(ECG_BUFFER_SIZE + 1)

    fun updateEcg(voltage: Int) {
        ecgSamplesList.addLast(EcgSample(voltage))
        if (ecgSamplesList.size > ECG_BUFFER_SIZE) ecgSamplesList.removeFirst()
        _ecgState.value = EcgState(
            ecgSamples = ecgSamplesList.toList(),
            currentVoltage = voltage
        )
    }

    fun clear() {
        ecgSamplesList.clear()
        _ecgState.value = EcgState()
    }
}

File: examples/example-ios/polar-sensor-data-collector/PSDC/Views/Graph/EcgDataHolder.swift
//  Copyright © 2026 Polar. All rights reserved.

import Foundation

struct EcgSamplePoint {
    let voltage: Int32
}

struct EcgState {
    var ecgSamples: [EcgSamplePoint] = []
    var currentVoltage: Int32 = 0
}

@MainActor
class EcgDataHolder: ObservableObject {
    static let shared = EcgDataHolder()

    private static let maxEcgSamples = 1300

    @Published private(set) var ecgState = EcgState()

    private var ecgSamplesList: [EcgSamplePoint] = []

    private init() {}

    func updateEcg(voltage: Int32) {
        let sample = EcgSamplePoint(voltage: voltage)
        ecgSamplesList.append(sample)

        if ecgSamplesList.count > Self.maxEcgSamples {
            ecgSamplesList.removeFirst(ecgSamplesList.count - Self.maxEcgSamples)
        }

        ecgState = EcgState(
            ecgSamples: ecgSamplesList,
            currentVoltage: voltage
        )
    }

    func clear() {
        ecgSamplesList.removeAll()
        ecgState = EcgState()
    }
}

